File: 0234-palindrome-linked-list/0234-palindrome-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
typedef struct ListNode* NODE;
NODE revers(NODE head){
    NODE temp=head;
    NODE p=0;
    NODE q=head;
    while(q){
        q=q->next;
        temp->next=p;
        p=temp;
        temp=q;
    }
head=p;
return p;
}
bool isPalindrome(struct ListNode* head) {
    NODE fast=head;
    NODE slow=head;
    while(fast!=NULL &&fast->next!=NULL){
        fast=fast->next->next;//move to two places ahead;
        slow=slow->next;//move to one place ahead;
    }
    slow=revers(slow);
    fast=head;
    while(slow){
        if(fast->val!=slow->val) return false;
        fast=fast->next;
        slow=slow->next;
    }
    return true;
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rev(ListNode l){
        ListNode p=null;
        ListNode q=l;
        while(q!=null){
            ListNode t=q.next;
            q.next=p;
            p=q;
            q=t;
        }
        return p;
        
    }
    public boolean isPalindrome(ListNode head) {
        ListNode l=head;
        ListNode r=head;
        ListNode q=head;
        while(r!=null &&r.next!=null){
           l=l.next;
           r=r.next.next;
        }
        if(r!=null){
            l=l.next;
        }
      
       
        ListNode p=rev(l);
       
        while(p!=null){
            if(q.val!=p.val)return false;
            p=p.next;
            q=q.next;
        }
return true;
    }
}